File: java/src/main/java/com/example/cote/PGMRS/보석_쇼핑.java
package com.example.cote.PGMRS;


import java.util.*;
import java.nio.file.*;
import java.io.*;

public class 보석_쇼핑 {

	private static class Range {
		public final int start;
		public final int end;
		public final int length;

		Range(int start, int end){
			this.start = start;
			this.end = end; 
			this.length = end - start; 
		}
	}

	private static int[] solve(String[] gems){

		Set<String> gemSet = new HashSet<>(List.of(gems));
		Map<String, Integer> includes = new HashMap<>();
		
		Range cur = new Range(0,1);
		Range answer = new Range(0, gems.length);
		includes.put(gems[0], 1);

		while(cur.start < gems.length){
			if(includes.keySet().size() == gemSet.size()){
				if(cur.length < answer.length){
					answer = cur;
				}

				includes.put(gems[cur.start], includes.get(gems[cur.start])-1);

				if(includes.get(gems[cur.start]) == 0){
					includes.remove(gems[cur.start]);
				}

				cur = new Range(cur.start+1, cur.end);
			}else if(cur.end < gems.length) {
				cur = new Range(cur.start, cur.end + 1);
				includes.put(gems[cur.end-1], includes.getOrDefault(gems[cur.end-1], 0)+1);
			}else {
				break;
			}
		}

		return new int[]{answer.start+1, answer.end};
	}
	
	public static void main(String... args)throws IOException{

		Path p = Paths.get(System.getProperty("user.dir") + "/data/보석_쇼핑.txt");
		BufferedReader rd = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(rd.readLine());

		while(testCases > 0){
			String[] gems = rd.readLine().split(" ");
			int[] result = Arrays.stream(rd.readLine().split(" "))
								.mapToInt(Integer::parseInt)
								.toArray();


			System.out.println("Solved : " + Arrays.toString(solve(gems)));
			System.out.println("Answer : " + Arrays.toString(result));

			testCases--;
		}

	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/보석_쇼핑.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*
import java.util.*

data class Range(val start: Int, val end: Int) {
	val length
		get() = end - start
}


fun solve(gems: List<String>): Range {

	val gemSet = setOf(*gems.distinct().toTypedArray())
	val includes = mutableMapOf<String,Int>()

	var answer = Range(0, gems.size)
	var cur = Range(0,1)

	includes[gems[0]] = 1

	while (cur.start < gems.size) {
		when {
			(includes.size == gemSet.size) -> {
				if (cur.length < answer.length) {
					answer = cur
				}

				includes[gems[cur.start]] = (includes[gems[cur.start]] ?: 1) -1

				if (includes[gems[cur.start]] == 0) {
					includes.remove(gems[cur.start])
				}

				cur = cur.copy(start=cur.start+1)
			}
			(cur.end < gems.size) -> {
				cur = cur.copy(end=cur.end+1)
				includes[gems[cur.end-1]] = (includes[gems[cur.end-1]] ?: 0) + 1 
			}
			else -> {
				break
			}
		}
	}


	return answer.copy(start = answer.start+1)
}

fun main() {

	val p = Paths.get(System.getProperty("user.dir")+"/data/보석_쇼핑.txt")
	val reader = Files.newBufferedReader(p)

	repeat(reader.readLine().toInt()) {
		val gems = reader.readLine().split(" ")
		val answer = reader.readLine()
						.split(" ")
						.map(String::toInt)

		val (start, end) = solve(gems)
		println(String.format("solved: (%d,%d),  answer: (%d,%d)", start, end, answer[0], answer[1]))
	}

	return
}